src/query/expression/src/type_check.rs

```diff
@@ -559,24 +558,24 @@ pub fn common_super_type(
                 .collect::<Option<Vec<_>>>()?;
             Some(DataType::Tuple(tys))
         }
-        (DataType::Number(_), DataType::Decimal(ty))
-        | (DataType::Decimal(ty), DataType::Number(_)) => {
-            let max_precision = ty.max_precision();
-            let scale = ty.scale();
-
-            DecimalDataType::from_size(DecimalSize {
-                precision: max_precision,
-                scale,
-            })
-            .ok()
-            .map(DataType::Decimal)
-        }
-
+        (DataType::String, decimal_ty @ DataType::Decimal(_))
+        | (decimal_ty @ DataType::Decimal(_), DataType::String) => Some(decimal_ty),
         (DataType::Decimal(a), DataType::Decimal(b)) => {
             let ty = DecimalDataType::binary_result_type(&a, &b, false, false, true).ok();
             ty.map(DataType::Decimal)
         }
-
+        (DataType::Number(num_ty), decimal_ty @ DataType::Decimal(_))
+        | (decimal_ty @ DataType::Decimal(_), DataType::Number(num_ty))
+            if !num_ty.is_float() =>
+        {
+            Some(decimal_ty)
+        }
+        (DataType::Number(num_ty), DataType::Decimal(_))
+        | (DataType::Decimal(_), DataType::Number(num_ty))
+            if num_ty.is_float() =>
```

Comment:
if `num_ty` is not float, should be decimal of `max_precision`.

Reply:
`range > range(num_ty)` is enough?

Reply:
> if num_ty is not float, should be decimal of max_precision.

if so, we should carefully compare the decimal in `can_auto_cast_to`: https://github.com/datafuselabs/databend/blob/45f3ca1bba14f1d192f9ab0094a9f63e9acfe792/src/query/expression/src/type_check.rs#L521

Reply:
But it's always true, because `10**38` > `2^64 - 1`

Reply:
I mean, for `(DataType::Number(n), DataType::Decimal(DecimalSize{precision: 1, scale: 1}))` , `can_auto_cast_to` should return false

Reply:
Yes!